### PR TITLE
ENH: Add a mypy plugin for exposing platform-specific extended-precision types

### PR DESCRIPTION
**Pull request by BvB93**
This PR modifies numpy's mypy plugin such that it manages all platform-specific extended-precision types (_e.g._ `float128`). More specifically, it _removes_ all extended-precision types unavailable to the platform in question, rather than adding them.

The advantage of their removal (rather than addition) is that aforementioned types will still be available for use even without enabling the plugin. On the flip side, this does mean that, without the plugin, every single extended-precision type is "available" on every single platform.

The plugin can be enabled via one's `mypy.ini` file:

``` ini
[mypy]
plugins = numpy.typing.mypy_plugin
```

Examples
----------
With the plugin enabled.
``` python
from typing import TYPE_CHECKING
import numpy as np

if TYPE_CHECKING:
    # Depending on the platform:
    # note: Revealed type is 'numpy.floating[numpy.typing._128Bit*]'
    # or 
    # error: Module has no attribute "float128"; maybe "float64", "float32", or "float16"?
    reveal_type(np.float128())
```

### CODE REVIEW

File: numpy/__init__.pyi [hunk 1]
```diff
@@ -128,6 +128,23 @@ from numpy.typing._callable import (
...
+    int128 as int128,
+    int256 as int256,
+    float80 as float80,
+    float96 as float96,
+    float128 as float128,
+    float256 as float256,
+    complex160 as complex160,
+    complex192 as complex192,
+    complex256 as complex256,
+    complex512 as complex512,
```
Comment (BvB93) [inline numpy/__init__.pyi:145]: Effectively the plugin removes all import statements that are invalid for the platform in question.

The full list of (potential) extended-precision types was taken from [`numpy/__init__.pxd`](https://github.com/numpy/numpy/blob/9f12028b9453c17b72b26355fd503e512af96a5d/numpy/__init__.pxd#L54-L79).

**Comment by BvB93** [comment]
Anyone knows what's going with the PyPy tests?
The failure seems to be unrelated, but it's failing nevertheless.

**Comment by mattip** [comment]
> Anyone knows what's going with the PyPy tests?

It uses a nightly build since there is not an official release for win64 yet. Let's see if this repeats.

**Comment by charris** [comment]
Thanks Bas.

**Pull request merged**